modules/040-node-manager/images/bashible-apiserver/pkg/template/steps_storage.go

```diff
@@ -17,48 +17,195 @@ limitations under the License.
+
+	configurationsChanged chan struct{}
+	emitter               changesEmitter
 }
 
-// NewStepsStorage creates StepsStorage for target and cloud provider.
-//
-// provider = aws | gcp | openstack | ...
-// target   = all | node-group
-func NewStepsStorage(rootDir, provider, target string) *StepsStorage {
-	config := config{rootDir, provider, target}
-	return &StepsStorage{config}
+type changesEmitter struct {
+	sync.Mutex
+	count int
 }
 
-func (s *StepsStorage) Render(templateContext map[string]interface{}) (map[string]string, error) {
-	bundle, ok := templateContext["bundle"].(string)
-	if !ok {
-		panic("expected string in templateContext[\"bundle\"]")
+// NewStepsStorage creates StepsStorage for target and cloud provider.
+func NewStepsStorage(ctx context.Context, rootDir string, ngConfigFactory dynamicinformer.DynamicSharedInformerFactory) *StepsStorage {
+	ss := &StepsStorage{
+		rootDir:                 rootDir,
+		systemScripts:           make(map[string]map[string][]byte),
+		nodeGroupConfigurations: make(map[string][]*nodeConfigurationScript),
+		configurationsChanged:   make(chan struct{}, 1),
 	}
 
-	templates, err := s.readBundleTemplates(bundle)
+	ss.subscribeOnCRD(ctx, ngConfigFactory)
+	return ss
+}
+
+func (s *StepsStorage) Render(target, bundle, provider string, templateContext map[string]interface{}, ng ...string) (map[string]string, error) {
+	steps, err := s.renderSystemScripts(target, bundle, provider, templateContext)
 	if err != nil {
 		return nil, err
 	}
 
-	steps := map[string]string{}
+	if len(ng) > 0 {
+		userConfigurations, err := s.renderNodeGroupConfigurations(bundle, ng[0], templateContext)
+		if err != nil {
+			klog.Errorf("Render user NodeGroupConfigurations failed: %s", err)
+			return steps, nil
+		}
+
+		for k, v := range userConfigurations {
+			if _, ok := steps[k]; ok {
+				klog.Errorf("NodeGroupConfigurations conflicts with system script: %s", k)
+				continue
+			}
+			steps[k] = v
+		}
+	}
+
+	return steps, nil
+}
+
+func (s *StepsStorage) OnNodeConfigurationsChanged() chan struct{} {
+	return s.configurationsChanged
+}
+
+func (s *StepsStorage) runBufferedEmitter() chan struct{} {
+	for {
+		time.Sleep(500 * time.Millisecond)
+		s.emitter.Lock()
+		if s.emitter.count > 0 {
+			s.configurationsChanged <- struct{}{}
+			s.emitter.count = 0
+		}
+		s.emitter.Unlock()
+	}
+}
+
+func (s *StepsStorage) emitChanges() {
+	s.emitter.Lock()
+	s.emitter.count++
+	s.emitter.Unlock()
+}
+
+func (s *StepsStorage) subscribeOnCRD(ctx context.Context, ngConfigFactory dynamicinformer.DynamicSharedInformerFactory) {
+	if ngConfigFactory == nil {
+		return
+	}
+
+	go s.runBufferedEmitter()
+
+	// Launch the informer
+	ginformer := ngConfigFactory.ForResource(schema.GroupVersionResource{
+		Group:    "deckhouse.io",
+		Version:  "v1alpha1",
+		Resource: "nodegroupconfigurations",
+	})
+
+	informer := ginformer.Informer()
+
+	informer.AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc: func(obj interface{}) {
```

Comment:
It is recommended to avoid writing logic right in the `AddFunc`, `UpdateFunc`, `DeleteFunc`. As far as I remember, it badly affects performance. Instead, you need to use some queue.

See this example from Prometheus https://github.com/prometheus/prometheus/blob/c0c22ed04200a8d24d1d5719f605c85710f0d008/discovery/kubernetes/pod.go#L60-L73

Reply:
Well. It makes sense on i/o operations. But I guest putting value into a memory map is not more complicated then put this value into a queue